### DISCUSSION TITLE
Load count of relation in populate?

### DISCUSSION THREAD

**Discussion by squivix** [mikro-orm/mikro-orm#4110] [Q&A]
I want to be able to  bulk load the count of a collection, possibly using `populate`.

For example if I have the following:

```ts
@Entity()
export class EntityA extends CustomBaseEntity {
    @ManyToMany({
        entity: () => EntityB,
        inversedBy: (entityB: EntityB) => entityB.entityAs,
        pivotEntity: () => PivotEntityAEntityB,
    })
    entityBs: Collection<EntityB> = new Collection<EntityB>(this);
}
```
I want to load a list of `EntityA`'s each with its count of `EntityB`'s:
```ts
const entityAs = await em.find(EntityA, {}, {populate: ["entityBs.count"]})
```

Right now the only ways I know of to achieve this are:

1. By populating the whole collection:
```ts
const entityAs = await em.find(EntityA, {}, {populate: ["entityBs"]})
```
2. With a formula:
```ts
@Entity()
export class EntityA extends CustomBaseEntity {
    @ManyToMany({
        entity: () => EntityB,
        inversedBy: (entityB: EntityB) => entityB.entityAs,
        pivotEntity: () => PivotEntityAEntityB,
    })
    entityBs: Collection<EntityB> = new Collection<EntityB>(this);

    @Formula((alias: string) => `(SELECT COUNT(entity_b_id) FROM pivot_entity_a_entity_b WHERE entity_a_id=${alias}.id)`, {
        type: "number"
    })
    entityBsCount!: number;
}
```
4. With `loadCount` called n times:
```ts
const entityAs = await em.find(EntityA, {})
await Promise.all(entityAs.map(async e => await e.entityBs.loadCount()));
```

None of which are very good. I don't need the whole collection, formulas are SQL-specific and don't get refreshed when I call `em.refresh` on the entity, `loadCount` n times is not ideal.

If there's another way I don't know if please let me know.

**Comment by B4nan** [comment]
I'd say the formula way is the right one, at least with the current state of things. You can also mark it as lazy so it's not selected by default (and use `populate` hint to select it). 

Or you can always store the count in a regular scalar property and maintain the value, possibly via `onFlush` event handler so you can do it from a single place. Of course this approach will be prone to errors if you touch the database manually too.

> formulas are SQL-specific

I don't think there is a better way with NoSQL drivers than the N+1 version with multiple `loadCount` calls.

>  don't get refreshed when I call em.refresh on the entity

Sounds like something to fix.

**Reply by squivix** [reply]
> at least with the current state of things. 

Are there any plans for a `loadCount` style function that works for an array of entities and doesn''t result in N+1 database calls?

**Reply by B4nan** [reply]
Maybe, but I think it would end up working the same as the formula with subquery. Granted, with much less boilerplate or the raw queries.
